### PR TITLE
Modify apple authenticator to enable federation flow in API based Authentication and Nonce validation

### PR DESCRIPTION
**Pull request by ZiyamSanthosh**
This PR includes the below given changes
- Enables API based authentication with federation flow for Apple authenticator. This supports both Native mode and Redirection mode.
- Enable nonce validation for apple authenticator.

Related issue:
- https://github.com/wso2/product-is/issues/20584

Merge the PR after merging the below given PR and bumping the oidc version:
- https://github.com/wso2-extensions/identity-outbound-auth-oidc/pull/186

**Sample Responses**

1. Initial `/authorize` request
```
{
    "flowId": "44f86b73-55fd-4ec5-8571-c7f4049d3c07",
    "flowStatus": "INCOMPLETE",
    "flowType": "AUTHENTICATION",
    "nextStep": {
        "stepType": "MULTI_OPTIONS_PROMPT",
        "authenticators": [
            {
                "authenticatorId": "R29vZ2xlT0lEQ0F1dGhlbnRpY2F0b3I6R29vZ2xl",
                "authenticator": "Google",
                "idp": "Google",
                "metadata": {
                    "i18nKey": "authenticator.google"
                }
            },
            {
                "authenticatorId": "QXBwbGVPSURDQXV0aGVudGljYXRvcjpBcHBsZQ",
                "authenticator": "Apple",
                "idp": "Apple",
                "metadata": {
                    "i18nKey": "authenticator.apple"
                }
            }
        ]
    },
    "links": [
        {
            "name": "authentication",
            "href": "https://wso2is.com:9443/oauth2/authn",
            "method": "POST"
        }
    ]
}
```

2. `/authn` request (Native mode)
```
{
    "flowId": "44f86b73-55fd-4ec5-8571-c7f4049d3c07",
    "flowStatus": "INCOMPLETE",
    "flowType": "AUTHENTICATION",
    "nextStep": {
        "stepType": "AUTHENTICATOR_PROMPT",
        "authenticators": [
            {
                "authenticatorId": "QXBwbGVPSURDQXV0aGVudGljYXRvcjpBcHBsZQ",
                "authenticator": "Apple",
                "idp": "Apple",
                "metadata": {
                    "i18nKey": "authenticator.apple",
                    "promptType": "INTERNAL_PROMPT",
                    "additionalData": {
                        "clientId": "com.asgupgrade.stage",
                        "scope": "email name"
                    }
                },
                "requiredParams": [
                    "accessToken",
                    "idToken"
                ]
            }
        ]
    },
    "links": [
        {
            "name": "authentication",
            "href": "https://wso2is.com:9443/oauth2/authn",
            "method": "POST"
        }
    ]
}
```

3. `/authn` request (Redirection mode)
```
{
    "flowId": "44f86b73-55fd-4ec5-8571-c7f4049d3c07",
    "flowStatus": "INCOMPLETE",
    "flowType": "AUTHENTICATION",
    "nextStep": {
        "stepType": "AUTHENTICATOR_PROMPT",
        "authenticators": [
            {
                "authenticatorId": "QXBwbGVPSURDQXV0aGVudGljYXRvcjpBcHBsZQ",
                "authenticator": "Apple",
                "idp": "Apple",
                "metadata": {
                    "i18nKey": "authenticator.apple",
                    "promptType": "REDIRECTION_PROMPT",
                    "additionalData": {
                        "state": "43fa600f-0ad3-4170-80e8-d8e615d59e09,OIDC",
                        "redirectUrl": "https://appleid.apple.com/auth/authorize?response_type=code&state=43fa600f-0ad3-4170-80e8-d8e615d59e09%2COIDC&redirect_uri=https%3A%2F%2Fgoogle.com&client_id=com.asgupgrade.stage&scope=email%20name&response_mode=form_post"
                    }
                },
                "requiredParams": [
                    "code",
                    "state"
                ]
            }
        ]
    },
    "links": [
        {
            "name": "authentication",
            "href": "https://wso2is.com:9443/oauth2/authn",
            "method": "POST"
        }
    ]
}
```

Related links:
[1] https://developer.apple.com/documentation/authenticationservices/asauthorizationopenidrequest/nonce

Sample JWT Payload from apple IDP
```
{
  "iss": "https://appleid.apple.com",
  "aud": "com.asgupgrade.stage",
  "exp": 1720100111,
  "iat": 1720013711,
  "sub": "001094.77fd79fb64a1413398438254b1a28e64.0535",
  "nonce": "f9945071-859e-4d2a-a6e4-5b435f16d109",
  "at_hash": "LfD82O14Te81Z_jqY6fAuQ",
  "email": "ziyam@wso2.com",
  "email_verified": true,
  "auth_time": 1720013689,
  "nonce_supported": true
}
```

### CODE REVIEW

**Review by Thumimku — APPROVED** [review]
LGTM

File: components/org.wso2.carbon.identity.application.authenticator.apple/src/main/java/org/wso2/carbon/identity/application/authenticator/apple/AppleAuthenticator.java [hunk 1]
```diff
@@ -145,8 +151,13 @@ protected void initiateAuthenticationRequest(HttpServletRequest request, HttpSer
                 String clientId = authenticatorProperties.get(OIDCAuthenticatorConstants.CLIENT_ID);
                 String authorizationEP = getAuthorizationServerEndpoint(authenticatorProperties);
                 String callbackUrl = getCallbackUrl(authenticatorProperties);
-                String state = getStateParameter(context, authenticatorProperties);
+                if (Boolean.parseBoolean((String) context.getProperty(IS_API_BASED))) {
+                    callbackUrl = resolveCallBackURLForAPIBasedAuthFlow(context);
```
Comment (pamodaaw) [inline components/org.wso2.carbon.identity.application.authenticator.apple/src/main/java/org/wso2/carbon/identity/application/authenticator/apple/AppleAuthenticator.java:155]: can't we have line 153 as the else block of this if?

Comment (pamodaaw) [reply components/org.wso2.carbon.identity.application.authenticator.apple/src/main/java/org/wso2/carbon/identity/application/authenticator/apple/AppleAuthenticator.java:155]: Or we can have `getCallbackUrl` method in this class and define this logic there

File: components/org.wso2.carbon.identity.application.authenticator.apple/src/main/java/org/wso2/carbon/identity/application/authenticator/apple/AppleAuthenticator.java [hunk 2]
```diff
@@ -145,8 +151,13 @@ protected void initiateAuthenticationRequest(HttpServletRequest request, HttpSer
                 String clientId = authenticatorProperties.get(OIDCAuthenticatorConstants.CLIENT_ID);
                 String authorizationEP = getAuthorizationServerEndpoint(authenticatorProperties);
                 String callbackUrl = getCallbackUrl(authenticatorProperties);
-                String state = getStateParameter(context, authenticatorProperties);
+                if (Boolean.parseBoolean((String) context.getProperty(IS_API_BASED))) {
```
Comment (pamodaaw) [inline components/org.wso2.carbon.identity.application.authenticator.apple/src/main/java/org/wso2/carbon/identity/application/authenticator/apple/AppleAuthenticator.java:154]: when this `IS_API_BASED` property is set in the context?

File: components/org.wso2.carbon.identity.application.authenticator.apple/src/main/java/org/wso2/carbon/identity/application/authenticator/apple/AppleAuthenticator.java (hunk 1 above)
Comment (ZiyamSanthosh) [reply components/org.wso2.carbon.identity.application.authenticator.apple/src/main/java/org/wso2/carbon/identity/application/authenticator/apple/AppleAuthenticator.java]: Addressed in https://github.com/wso2-extensions/identity-outbound-auth-apple/pull/9/commits/9f2bc6ff1d342b6c4ceeec138af629a4634850a5

File: components/org.wso2.carbon.identity.application.authenticator.apple/src/main/java/org/wso2/carbon/identity/application/authenticator/apple/AppleAuthenticator.java (hunk 2 above)
Comment (ZiyamSanthosh) [reply components/org.wso2.carbon.identity.application.authenticator.apple/src/main/java/org/wso2/carbon/identity/application/authenticator/apple/AppleAuthenticator.java]: This is being set at the framework[1].

[1] https://github.com/wso2/carbon-identity-framework/blob/3782fee47c7c612d37f955c12c5f70317c020e4b/components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultRequestCoordinator.java#L791-L794

Comment (pamodaaw) [reply components/org.wso2.carbon.identity.application.authenticator.apple/src/main/java/org/wso2/carbon/identity/application/authenticator/apple/AppleAuthenticator.java]: So why can't we use the same constant defined in the framework?

Comment (ZiyamSanthosh) [reply components/org.wso2.carbon.identity.application.authenticator.apple/src/main/java/org/wso2/carbon/identity/application/authenticator/apple/AppleAuthenticator.java]: This concern was addressed with https://github.com/wso2-extensions/identity-outbound-auth-apple/pull/9/commits/9f2bc6ff1d342b6c4ceeec138af629a4634850a5

File: components/org.wso2.carbon.identity.application.authenticator.apple/src/main/java/org/wso2/carbon/identity/application/authenticator/apple/AppleAuthenticator.java [hunk 3]
```diff
@@ -636,6 +645,56 @@ protected String getComponentId() {
...
+     * This method is responsible to return the scopes as space separated string.
+     *
+     * @param authenticatorProperties   Authenticator properties.
+     * @return Scopes string
+     */
+    @Override
+    protected String getScope(Map<String, String> authenticatorProperties) {
+
+        String scopes = authenticatorProperties.get(IdentityApplicationConstants.Authenticator.OIDC.SCOPES);
+        return scopes.replace(",", " ").replace("+", " ");
```
Comment (janakamarasena) [inline components/org.wso2.carbon.identity.application.authenticator.apple/src/main/java/org/wso2/carbon/identity/application/authenticator/apple/AppleAuthenticator.java:680]: Lets add a null check here

Comment (janakamarasena) [reply components/org.wso2.carbon.identity.application.authenticator.apple/src/main/java/org/wso2/carbon/identity/application/authenticator/apple/AppleAuthenticator.java:680]: also lets take the separators to constants

File: components/org.wso2.carbon.identity.application.authenticator.apple/src/main/java/org/wso2/carbon/identity/application/authenticator/apple/AppleAuthenticator.java [hunk 4]
```diff
@@ -636,6 +645,56 @@ protected String getComponentId() {
...
+    }
+
+    /**
+     * This method is responsible for resolving the callback URL for either API based authentication flow or the
+     * usual flow.
+     *
+     * @param context Authentication context.
+     * @return Callback URL.
+     */
+    private String getCallBackURL(AuthenticationContext context, Map<String, String> authenticatorProperties) {
```
Comment (pamodaaw) [inline components/org.wso2.carbon.identity.application.authenticator.apple/src/main/java/org/wso2/carbon/identity/application/authenticator/apple/AppleAuthenticator.java:690]: https://github.com/wso2-extensions/identity-outbound-auth-oidc/blob/master/components/org.wso2.carbon.identity.application.authenticator.oidc/src/main/java/org/wso2/carbon/identity/application/authenticator/oidc/OpenIDConnectAuthenticator.java#L310

Can't we inherit this method from `OpenIDConnectAuthenticator`

Comment (ZiyamSanthosh) [reply components/org.wso2.carbon.identity.application.authenticator.apple/src/main/java/org/wso2/carbon/identity/application/authenticator/apple/AppleAuthenticator.java]: This was addressed in https://github.com/wso2-extensions/identity-outbound-auth-apple/pull/9/commits/1a9f83d2deac9d435d73d8dcd3e3f08d792db3bc

File: components/org.wso2.carbon.identity.application.authenticator.apple/src/main/java/org/wso2/carbon/identity/application/authenticator/apple/AppleAuthenticator.java (hunk 3 above)
Comment (ZiyamSanthosh) [reply components/org.wso2.carbon.identity.application.authenticator.apple/src/main/java/org/wso2/carbon/identity/application/authenticator/apple/AppleAuthenticator.java]: This was addressed in both https://github.com/wso2-extensions/identity-outbound-auth-apple/pull/9/commits/1a9f83d2deac9d435d73d8dcd3e3f08d792db3bc and https://github.com/wso2-extensions/identity-outbound-auth-apple/pull/9/commits/84cbc2f539b418033c4f1c3e4549a75862b4c3f2

**Review by pamodaaw — APPROVED** [review]

**Review by shanggeeth — APPROVED** [review]

**Pull request merged**